torch/_decomp/decompositions.py

```diff
@@ -2284,6 +2284,32 @@ def _index_add(
         return out.squeeze(0) if zero_dim else out.contiguous()
 
 
+@register_decomposition(aten.pad_sequence.default)
+@aten.pad_sequence.default.py_impl(DispatchKey.CompositeImplicitAutograd)
+def pad_sequence(
+    sequences, batch_first=False, padding_value=0.0
+):
+    torch._check(len(sequences) > 0, lambda: "received an empty list of sequences")
+    sequences_size = len(sequences)
+    max_size = sequences[0].size()
+    trailing_dims = max_size[1:]
+    max_len = max([x.size(0) for x in sequences])
+    if batch_first:
+        out_dims = (sequences_size, max_len)
+    else:
+        out_dims = (max_len, sequences_size)
+    out_dims = out_dims + trailing_dims
+    out = sequences[0].new_full(out_dims, padding_value)
+    for i in range(sequences_size):
+        currseq = sequences[i]
+        length_i = currseq.size(0)
+        if batch_first:
+            out[i, :length_i].copy_(currseq)
```

Comment:
in-place ops in decompositions are not kosher. The way to do this is to figure out by how much you need to pad every tensor, and concatenate a `new_full` of the right length to every input. At the end, you just stack all of them together.

Reply:
One thing to note, I tried both and inductor actually does a much better with the copies... 1 kernel vs `sequences_size` kernels. 

I think the kosher way to write this would be
```
row = aten.constant_pad_nd(curseq, (0, row_size - curseq.size(0)), pad_value)
out = aten.select_scatter(out, row, dim=dim, i)
```

Of course in an ideal world `torch.cat` would be able to code that's just as good.

Reply:
I was going to say that the generated code was going to look horrible, but note that this PR doesn't add this decomposition to core so that's alright.

I was thinking about doing something with `pad` but didn't post it either. Agree this is a much better way of doing this. We should try write some passes to go from full + cat -> pad and, if it generates better code, from stack into a bunch of select scatter (or just generate better code for cat directly...)